src/main/java/io/quarkus/gizmo/Type.java

```diff
@@ -0,0 +1,418 @@
+ * <p>
+ * Implementations are instantiated via factory methods; for example {@link #voidType()} and {@link #classType(Class)}.
+ * 
+ * @see SignatureBuilder
+ */
+public interface Type {
+
+    // factory methods
+
+    public static VoidType voidType() {
+        return VoidType.INSTANCE;
+    }
+
+    public static PrimitiveType byteType() {
+        return PrimitiveType.BYTE;
+    }
+
+    public static PrimitiveType booleanType() {
+        return PrimitiveType.BOOLEAN;
+    }
+
+    public static PrimitiveType intType() {
+        return PrimitiveType.INT;
+    }
+
+    public static PrimitiveType longType() {
+        return PrimitiveType.LONG;
+    }
+
+    public static PrimitiveType shortType() {
+        return PrimitiveType.SHORT;
+    }
+
+    public static PrimitiveType doubleType() {
+        return PrimitiveType.DOUBLE;
+    }
+
+    public static PrimitiveType floatType() {
+        return PrimitiveType.FLOAT;
+    }
+
+    public static PrimitiveType charType() {
+        return PrimitiveType.CHAR;
+    }
+
+    public static ClassType classType(DotName name) {
+        return new ClassType(Objects.requireNonNull(name).toString().replace('.', '/'));
+    }
+
+    public static ClassType classType(String name) {
+        return new ClassType(name);
+    }
+
+    public static ClassType classType(Class<?> classType) {
+        Objects.requireNonNull(classType);
+        if (classType.getDeclaringClass() != null) {
+            return classType(classType.getDeclaringClass().getName().replace('.', '/') + "." + classType.getSimpleName());
+        }
+        return classType(classType.getName().replace('.', '/'));
+    }
+
+    public static ParameterizedType parameterizedType(ClassType classType, Type... typeArguments) {
+        return new ParameterizedType(classType, Arrays.asList(typeArguments));
+    }
+
+    public static ArrayType arrayType(Type component) {
+        return new ArrayType(component, 1);
+    }
+
+    public static ArrayType arrayType(Type component, int dimensions) {
+        return new ArrayType(component, dimensions);
+    }
+
+    public static TypeVariable typeVariable(String name) {
+        return typeVariable(name, ClassType.OBJECT);
+    }
+
+    public static TypeVariable typeVariable(String name, Type classBound, Type... interfaceBounds) {
+        return new TypeVariable(name, classBound, Arrays.asList(interfaceBounds));
+    }
+
+    public static WildcardType wildcardTypeWithUpperBound(Type bound) {
+        return new WildcardType(bound, null);
+    }
+
+    public static WildcardType wildcardTypeWithLowerBound(Type bound) {
+        return new WildcardType(null, bound);
+    }
+
+    public static WildcardType wildcardTypeUnbounded() {
+        return new WildcardType(ClassType.OBJECT, null);
+    }
+
+    /**
+     * 
+     * @return the signature as defined in JVMS 17, chapter "4.7.9.1. Signatures"
+     */
+    String toSignature();
```

Comment:
Instead of, or in addition to, a `String toSignature()` method, I'd probably use `void appendSignature(StringBuilder builder)`. Obtaining a signature of a type in isolation is generally useless, I'd say.

Reply:
Ok, we can have both.